core/drand_daemon.go

```diff
@@ -0,0 +1,178 @@
+
+	return drandDaemon, nil
+}
+
+
+func (dd *DrandDaemon) RemoteStatus(ctx context.Context, request *drand.RemoteStatusRequest) (*drand.RemoteStatusResponse, error) {
+	bp, _, err := dd.getBeaconProcess(request.Metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	return bp.RemoteStatus(ctx, request)
+}
+
+func (dd *DrandDaemon) init() error {
+	c := dd.opts
+
+	// Set the private API address to the command-line flag, if given.
+	// Otherwise, set it to the address associated with stored private key.
+	privAddr := c.PrivateListenAddress("")
+	pubAddr := c.PublicListenAddress("")
+
+	if privAddr == "" {
+		return fmt.Errorf("private listen address cannot be empty")
+	}
+
+	// ctx is used to create the gateway below.
+	// Gateway constructors (specifically, the generated gateway stubs that require it)
+	// do not actually use it, so we are passing a background context to be safe.
+	ctx := context.Background()
+
+	var err error
+	dd.log.Infow("", "network", "init", "insecure", c.insecure)
+
+	if pubAddr != "" {
+		handler, err := http.New(ctx, &drandProxy{dd}, c.Version(), dd.log.With("server", "http"))
+		if err != nil {
+			return err
+		}
+		if dd.pubGateway, err = net.NewRESTPublicGateway(ctx, pubAddr, c.certPath, c.keyPath, c.certmanager, handler, c.insecure); err != nil {
+			return err
+		}
+	}
+
+	dd.privGateway, err = net.NewGRPCPrivateGateway(ctx, privAddr, c.certPath, c.keyPath, c.certmanager, dd, c.insecure, c.grpcOpts...)
+	if err != nil {
+		return err
+	}
+
+	p := c.ControlPort()
+	dd.control = net.NewTCPGrpcControlListener(dd, p)
+	go dd.control.Start()
+
+	dd.log.Infow("", "private_listen", privAddr, "control_port", c.ControlPort(), "public_listen", pubAddr, "folder", c.ConfigFolderMB())
+	dd.privGateway.StartAll()
+	if dd.pubGateway != nil {
+		dd.pubGateway.StartAll()
+	}
+
+	return nil
+}
+
+//InstantiateBeaconProcess creates a new BeaconProcess linked to beacon with id 'beaconID'
+func (dd *DrandDaemon) InstantiateBeaconProcess(beaconID string, store key.Store) (*BeaconProcess, error) {
+	if beaconID == "" {
+		beaconID = common.DefaultBeaconID
+	}
+
+	bp, err := NewBeaconProcess(dd.log, dd.version, store, dd.opts, dd.privGateway, dd.pubGateway, dd.control)
+	if err != nil {
+		return nil, err
+	}
+
+	dd.state.Lock()
+	dd.beaconProcesses[beaconID] = bp
+	dd.state.Unlock()
+
+	return bp, nil
+}
+
+// LoadBeacons checks for existing stores and creates the corresponding BeaconProcess
+// accordingly to each stored BeaconID
+func (dd *DrandDaemon) LoadBeacons(metricsFlag string) error {
+	// Load possible existing stores
+	stores, err := key.NewFileStores(dd.opts.ConfigFolderMB())
+	if err != nil {
+		return err
+	}
+
+	for beaconID, fs := range stores {
+		bp, err := dd.InstantiateBeaconProcess(beaconID, fs)
+		if err != nil {
+			fmt.Printf("beacon id [%s]: can't instantiate randomness beacon. err: %s \n", beaconID, err)
+			return err
+		}
+
+		err, freshRun := bp.Load()
+		if err != nil {
+			return err
+		}
+
+		if freshRun {
```

Comment:
This logic should be done by the beacon process itself - daemon shouldn't care about the internal process. Let's decouple as much as possible the logic. so "Load()" should start beacon if it is needed or not by itself.

Reply:
Referenced in https://github.com/drand/drand/issues/860